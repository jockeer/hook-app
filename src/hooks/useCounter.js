import { useState } from "react";

const useCounter = ( initialState = 10 ) => {
    const [counter, setCounter] = useState(initialState)

    const increment = (factor = 1)=>{
        setCounter(counter + factor )
    }
    const descrement = (factor = 1)=>{
        setCounter(counter - factor )
    }
    const reset = ()=>{
        setCounter(initialState)
    }

    return {
        counter,
        increment,
        descrement,
        reset
    }
}
 
export default useCounter;