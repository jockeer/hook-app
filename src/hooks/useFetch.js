import { useEffect, useRef, useState } from "react";


const useFetch = ( url ) => {

    const isMounted = useRef(true); //! el useRef nos servira para verificar que el componente este montado para setear su estado
    
    const [state, setState] = useState({
        data: null, //!mejor trabajar con arreglos vacios 
        loading: true, 
        error:null 
    })

    useEffect(()=>{
        return () => {
            isMounted.current = false;
        }
    }, [])
    
    useEffect(()=>{

        setState({data:null, loading:true,error:null})

        fetch(url)
            .then( resp => resp.json())
            .then( data => {

                if (isMounted.current) {
                    setState({
                        loading: false,
                        error: null,
                        data
                    })
                }
            })
        
    }, [ url ])

    return state

}
 
export default useFetch;