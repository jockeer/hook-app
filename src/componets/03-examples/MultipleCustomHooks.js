import React from 'react';
import useCounter from '../../hooks/useCounter';
import useFetch from '../../hooks/useFetch';

import '../01-useState/counter.css'

const MultipleCustomHooks = () => {
    
    const { counter, increment } = useCounter(1) 

    const { loading, data } =  useFetch(`https://breakingbadapi.com/api/quotes/${counter}`)

    const { author, quote } = !!data && data[0] //!la doble negacion '!!' en la condicion vuelve el valor null en false
    
    return (  
        <div>
            <h1>BreakingBad Quotes</h1>
            <hr />

            { loading 
                ? 
                <div className="alert alert-info text-center">
                    Loading...
                </div>
                : 
                <blockquote className="blockquote text-end">
                    <p className="mb-3">{quote}</p>
                    <footer className="blockquote-footer">{author}</footer>
                </blockquote>
            }

            <button className='btn btn-primary m-2' onClick={() => increment(1)}>Siguiente Quote</button>

            
        </div>
    );
}
 
export default MultipleCustomHooks;