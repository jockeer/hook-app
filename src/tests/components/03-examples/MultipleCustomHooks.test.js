import React from 'react';
import '@testing-library/jest-dom'
import {shallow} from 'enzyme'
import MultipleCustomHooks from '../../../componets/03-examples/MultipleCustomHooks';
import useFetch from '../../../hooks/useFetch';
import useCounter from '../../../hooks/useCounter';

jest.mock('../../../hooks/useFetch')
jest.mock('../../../hooks/useCounter')




describe('Pruebas en MultipleCustomHooks', () => { 
    
    useCounter.mockReturnValue({
        counter:10,
        increment: ()=>{}
    })

    test('Debe mostrar correctamente', () => { 
        
        useFetch.mockReturnValue({
            data:null,
            loading:true,
            error:null
        })

        const wrapper = shallow(<MultipleCustomHooks />)
        expect(wrapper).toMatchSnapshot()
    })

    test('Debe mostrar la informacion', () => {
        useFetch.mockReturnValue({
            data:[{
                author:'Daniel',
                quote:'Hola mundo'
            }],
            loading:false,
            error:null
        })

        const wrapper = shallow(<MultipleCustomHooks />)

        expect(wrapper.find('.alert').exists()).toBe(false)
        expect(wrapper.find('.mb-3').text().trim()).toBe('Hola mundo')
        expect(wrapper.find('footer').text().trim()).toBe('Daniel')

        // console.log(wrapper.html())
        

    })

})