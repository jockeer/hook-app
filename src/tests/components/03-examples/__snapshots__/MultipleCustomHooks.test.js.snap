// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en MultipleCustomHooks Debe mostrar correctamente 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary m-2"
    onClick={[Function]}
  >
    Siguiente Quote
  </button>
</div>
`;
